boot/seal.go

```diff
@@ -25,61 +25,327 @@ import (
+// listed in modeenv.
+func runModeKernelsFromModeenv(modeenv *Modeenv) ([]string, error) {
+	switch len(modeenv.CurrentKernels) {
+	case 1:
+		current := filepath.Join(dirs.SnapBlobDir, modeenv.CurrentKernels[0])
+		return []string{current}, nil
+	case 2:
+		current := filepath.Join(dirs.SnapBlobDir, modeenv.CurrentKernels[0])
+		next := filepath.Join(dirs.SnapBlobDir, modeenv.CurrentKernels[1])
+		return []string{current, next}, nil
+	}
+	return nil, fmt.Errorf("invalid number of kernels in modeenv")
+}
+
+func sealKeyParams(pbc predictableBootChains) (*secboot.SealKeyParams, error) {
+	modelParams := make([]*secboot.SealKeyModelParams, 0, len(pbc))
+	for _, bc := range pbc {
+		loadChains, err := efiLoadChains(bc)
+		if err != nil {
+			return nil, fmt.Errorf("error building EFI load chains: %s", err)
+		}
+
+		modelParams = append(modelParams, &secboot.SealKeyModelParams{
+			Model:          bc.model,
+			KernelCmdlines: []string{bc.KernelCmdline},
+			EFILoadChains:  loadChains,
+		})
 	}
 
-	sealKeyParams := secboot.SealKeyParams{
-		ModelParams: []*secboot.SealKeyModelParams{
-			{
-				Model:          model,
-				KernelCmdlines: kernelCmdlines,
-				EFILoadChains:  [][]bootloader.BootFile{loadChain},
-			},
-		},
+	sealKeyParams := &secboot.SealKeyParams{
+		ModelParams:             modelParams,
 		KeyFile:                 filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
 		TPMPolicyUpdateDataFile: filepath.Join(InstallHostFDEDataDir, "policy-update-data"),
 		TPMLockoutAuthFile:      filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
 	}
 
-	if err := secbootSealKey(key, &sealKeyParams); err != nil {
-		return fmt.Errorf("cannot seal the encryption key: %v", err)
+	return sealKeyParams, nil
+}
+
+func efiLoadChains(bc bootChain) ([][]bootloader.BootFile, error) {
+	seq0 := make([]bootloader.BootFile, 0, len(bc.AssetChain)+1)
+	seq1 := make([]bootloader.BootFile, 0, len(bc.AssetChain)+1)
+
+	hasDifferentHashes := false
+
+	for _, ba := range bc.AssetChain {
+		if len(ba.Hashes) > 1 {
+			hasDifferentHashes = true
+		}
+		p0, p1, err := cachedAssetPathnames(bc.blName, ba.Name, ba.Hashes)
+		if err != nil {
+			return nil, err
+		}
+		seq0 = append(seq0, bootloader.NewBootFile("", p0, bootloader.Role(ba.Role)))
+		seq1 = append(seq1, bootloader.NewBootFile("", p1, bootloader.Role(ba.Role)))
 	}
 
-	return nil
+	// add kernel
+	seq0 = append(seq0, bc.kernelBootFile)
+	seq1 = append(seq1, bc.kernelBootFile)
+
+	// XXX: we can explode to all possible combinations now, or using load event trees later
+
+	if hasDifferentHashes {
+		return [][]bootloader.BootFile{seq0, seq1}, nil
+	}
+
+	return [][]bootloader.BootFile{seq0}, nil
+
+}
+
+// cachedAssetPathnames returns the pathnames of the files corresponding
+// to the current and next instances of a given boot asset.
+func cachedAssetPathnames(blName, name string, hashes []string) (current, next string, err error) {
```

Comment:
this is a strange function (maybe it comes from older code you had around ? ),  I think we want a variant that takes only one hash, also this function should probably live in assets.go itself

Reply:
This function is no longer necessary after the changes in PR #9306 and was removed.